pkg/cdi/cdi.go

```diff
@@ -0,0 +1,105 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdi
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	cdiSpecs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"github.com/golang/glog"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+// CreateCDISpec creates CDI spec file with specified devices
+func CreateCDISpec(resourcePrefix string, filteredDevices []types.HostDevice, rPool types.ResourcePool) error {
+	prefix := resourcePrefix
+	if prefixOverride := rPool.GetResourcePrefix(); prefixOverride != "" {
+		prefix = prefixOverride
+	}
+	cdiDevices := make([]cdiSpecs.Device, 0)
+	cdiSpec := cdiSpecs.Spec{
+		Version: cdiSpecs.CurrentVersion,
+		Kind:    prefix + "/net",
+		Devices: cdiDevices,
+	}
+
+	for _, dev := range filteredDevices {
+		containerEdit := cdiSpecs.ContainerEdits{
+			DeviceNodes: make([]*cdiSpecs.DeviceNode, 0),
+			Env:         make([]string, 0),
+		}
+
+		envs, err := rPool.GetEnvs(prefix, []string{dev.GetDeviceID()})
+		if err != nil {
+			glog.Errorf("CreateCDISpec(): Unable to create environment variables: %s", err)
+			return fmt.Errorf("unable to create environment variables: %s", err)
+		}
+		for name, value := range envs {
+			containerEdit.Env = append(containerEdit.Env, name+"="+value)
+		}
+
+		for i, spec := range dev.GetDeviceSpecs() {
+			glog.Infof("Checking device spec (%d): %s", i, spec.String())
+			deviceNode := cdiSpecs.DeviceNode{
+				Path:        spec.ContainerPath,
+				HostPath:    spec.HostPath,
+				Permissions: "rwm",
+			}
+			containerEdit.DeviceNodes = append(containerEdit.DeviceNodes, &deviceNode)
+		}
+		device := cdiSpecs.Device{
+			Name:           dev.GetDeviceID(),
+			ContainerEdits: containerEdit,
+		}
+		cdiSpec.Devices = append(cdiSpec.Devices, device)
+	}
+
+	err := cdi.GetRegistry().SpecDB().WriteSpec(&cdiSpec, prefix)
+	if err != nil {
+		glog.Errorf("CreateCDISpec(): can not create CDI json: %q", err)
+		return err
+	}
+
+	return nil
+}
+
+// CreateContainerAnnotations creates container annotations based on CDI spec for a container runtime
+func CreateContainerAnnotations(devicesIDs []string, cdiCache *cdi.Cache) (map[string]string, error) {
+	annotations := make(map[string]string, 0)
+	for _, id := range devicesIDs {
+		var cdiDevice *cdi.Device
+
+		for _, device := range cdiCache.ListDevices() {
+			if strings.Contains(device, id) {
+				cdiDevice = cdiCache.GetDevice(device)
+			}
+		}
+		if cdiDevice != nil {
+			// TODO: (e0ne): add support of several devices
+			annotations["cdi.k8s.io/"+strings.ReplaceAll(cdiDevice.GetSpec().Kind, "/", "_")] = cdiDevice.GetSpec().Kind + "=" + cdiDevice.Name
```

Comment:
can you provide more info on the expected format of this annotation expected by container runtime ?

Reply:
refactored to use existing function